### PR TITLE
Add support to cross-compile architectures.

### PR DESCRIPTION
**Pull request by djlwilder**
Add support to cross-compile architectures.
Signed-off-by: David Wilder <wilder@us.ibm.com>

##  Description
* Adds support for cross-building images.
* Adds native build support for arm64 and s390x.
* Adds standard _build_ and _image_ targets per @caseydavenport.
* Standardizes Dockerfile names per the other cross-build projects.
* Pin go-build to V0.14. (will need updating, see note below)
* Added _help_ make target.
* Removed _bin/docker_ target, no-longer necessary as docker has been added into go-build.

To build native: 
	**make image**  or **make all**

To cross build (when running on amd64 only):
	**ARCH=arch make image**
	arch is one of: **ppc64le, arm64 or s390x**.

Note: Depends on https://github.com/projectcalico/go-build/pull/42 for running tests on ppc64le and s390x.

### CODE REVIEW

**Comment by djlwilder** [comment]
@deitch

**Comment by deitch** [comment]
This is nice @djlwilder . I only have a few comments. I will put them inline.

File: Makefile [hunk 1]
```diff
@@ -28,17 +60,43 @@ LOCAL_IP_ENV?=$(shell ip route get 8.8.8.8 | head -1 |  awk '{print $$7}')
 
 # Can choose different docker versions see list here - https://hub.docker.com/_/docker/
 HOST_CHECKOUT_DIR?=$(CURDIR)
-CONTAINER_NAME?=calico/libnetwork-plugin$(ARCHTAG)
-GO_BUILD_CONTAINER?=calico/go-build$(ARCHTAG):$(GO_BUILD_VER)
+CONTAINER_NAME?=calico/libnetwork-plugin:latest-$(ARCH)
 DIST=dist/$(ARCH)
 PLUGIN_LOCATION?=$(CURDIR)/$(DIST)/libnetwork-plugin
```
Comment (deitch) [inline Makefile]: Consistent naming: In the other builds, we have used a single `dist/` or `bin/` directory (probably should unify that too, but doesn't matter all that much), with the output files named `dist/<filename>-<arch>`, e.g. `felix-amd64` or `typha-ppc64le`, whereas this one does `dist/<arch>/<filename>`, e.g. `dist/amd64/libnetwork-plugin`. 

I think we should take advantage of this cleanup to unify the filename, make the outputs `dist/libnetwork-plugin-<arch>`.

It would mean changes to the `Makefile` and the `Dockerfile.*`, but that is it.

File: Makefile [hunk 2]
```diff
@@ -67,14 +126,17 @@ $(DIST)/libnetwork-plugin: vendor
 		-e ARCH=$(ARCH) \
 		$(GO_BUILD_CONTAINER) sh -c '\
 			cd /go/src/github.com/projectcalico/libnetwork-plugin && \
-			make build'
+			make binary'
 
-build: $(SRC_FILES) vendor
-	CGO_ENABLED=0 go build -v -i -o $(DIST)/libnetwork-plugin -ldflags "-X main.VERSION=$(shell git describe --tags --dirty) -s -w" main.go
+binary: $(SRC_FILES) vendor
+	CGO_ENABLED=0 GOARCH=$(ARCH) go build -v -i -o $(DIST)/libnetwork-plugin -ldflags "-X main.VERSION=$(shell git describe --tags --dirty) -s -w" main.go
```
Comment (deitch) [inline Makefile]: Release default: If you look at the [releases page](https://github.com/projectcalico/libnetwork-plugin/releases), you see that the unnamed `libnetwork-plugin` is assumed to be `amd64`. I think we need to have in the `dist/` dir, in addition to `libnetwork-plugin-<arch>`, a single file `libnetwork-plugin` which is `amd64`. As @caseydavenport pointed out, we eventually should change our documentation and other processes to eliminate that, but it is _far_ easier to add all of the archs as part of the release and a default "archless" file which is just `cp libnetwork-plugin-amd64 libnetwork-plugin`, and then later change the dependent docs and installs. 

The above just needs something like:

```make
ifeq ($(ARCH),amd64)
     cp dist/libnetwork-plugin-amd64 dist/libnetwork-plugin
endif
```

File: Makefile [hunk 3]
```diff
@@ -104,20 +166,20 @@ endif
 	# Check that the version output appears on a line of its own (the -x option to grep).
 	# Tests that the "git tag" makes it into the binary. Main point is to catch "-dirty" builds
 	@echo "Checking if the tag made it into the binary"
-	docker run --rm calico/libnetwork-plugin$(ARCHTAG) -v | grep -x $(VERSION) || (echo "Reported version:" `docker run --rm calico/libnetwork-plugin$(ARCHTAG) -v` "\nExpected version: $(VERSION)" && exit 1)
-	docker tag calico/libnetwork-plugin$(ARCHTAG) calico/libnetwork-plugin$(ARCHTAG):$(VERSION)
-	docker tag calico/libnetwork-plugin$(ARCHTAG) quay.io/calico/libnetwork-plugin$(ARCHTAG):$(VERSION)
-	docker tag calico/libnetwork-plugin$(ARCHTAG) quay.io/calico/libnetwork-plugin$(ARCHTAG):latest
+	docker run --rm $(CONTAINER_NAME) -v | grep -x $(VERSION) || (echo "Reported version:" `docker run --rm $(CONTAINER_NAME) -v` "\nExpected version: $(VERSION)" && exit 1)
+	docker tag $(CONTAINER_NAME) calico/libnetwork-plugin:$(VERSION)
```
Comment (deitch) [inline Makefile]: This should be `calico/libnetwork-plugin:$(VERSION)-$(ARCH)`, as should the next 2 lines have `-$(ARCH)`

File: Makefile [hunk 4]
```diff
@@ -104,20 +166,20 @@ endif
...
+	docker tag $(CONTAINER_NAME) quay.io/calico/libnetwork-plugin:$(VERSION)
+	docker tag $(CONTAINER_NAME) quay.io/calico/libnetwork-plugin:latest
 
 	@echo "Now push the tag and images. Then create a release on Github and attach the $(DIST)/libnetwork-plugin binary"
 	@echo "git push origin $(VERSION)"
-	@echo "docker push calico/libnetwork-plugin$(ARCHTAG):$(VERSION)"
-	@echo "docker push quay.io/calico/libnetwork-plugin$(ARCHTAG):$(VERSION)"
-	@echo "docker push calico/libnetwork-plugin$(ARCHTAG):latest"
-	@echo "docker push quay.io/calico/libnetwork-plugin$(ARCHTAG):latest"
+	@echo "docker push calico/libnetwork-plugin:$(VERSION)"
```
Comment (deitch) [inline Makefile]: These need `-$(ARCH)` as well.

File: Makefile (hunk 3 above)
Comment (deitch) [reply Makefile]: These need to add the "arch-less" tags with `amd64` as default. See [here](https://github.com/projectcalico/typha/blob/master/Makefile#L342) and [here](https://github.com/projectcalico/typha/blob/master/Makefile#L387) and [here](https://github.com/projectcalico/typha/blob/master/Makefile#L391) (and following lines) from the typha `Makefile`

**Comment by djlwilder** [comment]
Updated  based on @deitch review comments (thanks for the review).
* Switched destination from dist/$(ARCH)/libnetwork-plugin  to dist/libnetwork-plugin-$(ARCH)
* When ARCH=amd64 dist/libnetwork-plugin-amd64 is copied to dist/libnetwork-plugin.
* Made suggested changes to the release target.

Tests run:
On amd64 hardware I ran:  
_make all,  ARCH=arm64 make all, ARCH=ppc64le make all, ARCH=s390x make all_

On ppc64le hardware I ran:
_make all_

File: Makefile [hunk 5]
```diff
@@ -100,26 +164,36 @@ ifndef VERSION
...
+ifeq ($(ARCH),amd64)
+	@echo "docker push calico/libnetwork-plugin:$(VERSION)"
+	@echo "docker push quay.io/calico/libnetwork-plugin:$(VERSION)"
+	@echo "docker push calico/libnetwork-plugin:latest"
+	@echo "docker push quay.io/calico/libnetwork-plugin:latest"
+endif
+	@echo "docker push calico/libnetwork-plugin:$(VERSION)-$(ARCH)"
+	@echo "docker push quay.io/calico/libnetwork-plugin:$(VERSION)-$(ARCH)"
+	@echo "docker push calico/libnetwork-plugin:latest-$(ARCH)"
+	@echo "docker push quay.io/calico/libnetwork-plugin:latest-$(ARCH)"
```
Comment (mkumatag) [inline Makefile]: can we create a target here which pushes images and mention that make target here so that it will become easy whovewer releasing them? someting like `make push`

Comment (djlwilder) [reply Makefile]: Adding a push target that that will actually do the push sounds good to me,  but as I wont be doing the release I would like input from the maintainer.  @robbrockbank

Comment (deitch) [reply Makefile]: I had a long discussion on Slack with @fasaxc about target `push`. As I understand it, there is a define desire to do a `push` target.

I captured the desired semantics as I understood them in [felix PR](https://github.com/projectcalico/felix/pull/1746) but it hasn't been merged yet, so still might change. Specifically, look [here](https://github.com/projectcalico/felix/pull/1746/files#diff-b67911656ef5d18c4ae36cb6741b7965R304)

```Makefile
push: image imagetag
	docker tag calico/felix:latest-$(ARCH) calico/felix:$(IMAGETAG)-$(ARCH)
	docker tag calico/felix:latest-$(ARCH) quay.io/calico/felix:$(IMAGETAG)-$(ARCH)
	docker push calico/felix:$(IMAGETAG)-$(ARCH)
	docker push quay.io/calico/felix:$(IMAGETAG)-$(ARCH)
ifeq ($(ARCH),amd64)
	docker tag calico/felix:latest-$(ARCH) calico/felix:$(IMAGETAG)
	docker tag calico/felix:latest-$(ARCH) quay.io/calico/felix:$(IMAGETAG)
	docker push calico/felix:$(IMAGETAG)
	docker push quay.io/calico/felix:$(IMAGETAG)
endif
```

Basically `make push`:

* checks that `IMAGETAG` is defined
* tags and pushes the current `:IMAGETAG-ARCH` for docker hub and quay.io
* if `amd64`, tags and pushes out the archless `:IMAGETAG` for docker hub and quay.io

The last step will be changed for multi-arch manifest if/when product management decide it is good.

Comment (deitch) [reply Makefile]: Sorry to drive you a little crazy with the changes. We now have the (modified) version of how `push` etc. work in `felix`, `kube-controllers`, etc. I will reference below.

**Comment by deitch** [comment]
Most of this looks pretty good. I would do one last run of changes to the `Makefile`. FYI, I will be working on `calico/node` to complete it and get it cross-arch once this (and a few other PRs) are in.

@djlwilder here is what we have in the `felix` (and other projects) as standard:

```Makefile
```

**Comment by djlwilder** [comment]
@deitch.
Sorry for the long delay in responding,  life is taking me in different directions :)
I updated the Makefile adding the tag and push targets as suggested. I attempted to follow the format used in the kube-controllers.  Thanks for your understanding.

**Comment by deitch** [comment]
Hey @djlwilder this is looking good. I am sorry you have to go through all of this. You are putting into a single PR work we did in 2-3 PRs in the other projects, so almost a moving target.

Two comments:

1. If you look at the Semaphore target, it runs by doing `make calico/libnetwork-plugin`. This no longer has such a target, so it needs to be added. It used to have it [here](https://github.com/projectcalico/libnetwork-plugin/pull/176/files#diff-b67911656ef5d18c4ae36cb6741b7965L76) but now it is gone. What you did is the right thing, but we need to do it in 3 phases: first, add all of these in while ensuring CI still works; second change CI to `make image`; third, remove the old target.
2. You wrote:

> When ARCH=amd64 dist/libnetwork-plugin-amd64 is copied to dist/libnetwork-plugin

per the earlier review. I couldn't figure out where that is in the `Makefile`?

**Comment by mkumatag** [comment]
@djlwilder can you please fix the comments?

**Comment by djlwilder** [comment]
Added back calico/libnetwork-plugin make target per review comments.

I miss spoke when I said "dist/libnetwork-plugin-amd64 is copied to dist/libnetwork-plugin".  I think I meant to say  "amd64 builds are re-tagged as :latest because Semaphore depends on that.
   

> docker tag calico/libnetwork-plugin:latest-amd64 calico/libnetwork-plugin:latest

If Ci has a dependency directly on dist/libnetwork-plugin let me know.

**Comment by deitch** [comment]
Looks like CI passes now. Here are the dependencies of (current) Semaphore CI:

* `make test-containerized`
* `make calico/libnetwork-plugin`
* image named `calico/libnetwork-plugin`

Once this is merged in, one of the team can change the 2nd dependency from `make calico/libnetwork-plugin` to `make image`, and we can eliminate that target.

We will need the `ci` and `cd` targets, but let's leave that for a separate PR. We have done about a dozen of just those already.

@fasaxc @tomdee @caseydavenport this looks good. Can we get it merged in?
